### ISSUE TITLE
Issue with calculation of Water Year

### ISSUE THREAD

**Issue by jenskiesel** [DOI-USGS/EflowStats#196]
The "validate_data" function does not work as expected for leap years, if yearType="water" and if users set "wyMonth=1" or "wyMonth=2". 
I think the underlying issue is, that the "get_waterYear" does not return the 'correct' water year associated to each day of the year if the wyMonth is Jan or Feb (all works as expected if the water year is set to start in March or later). The problem is, that the days in Jan+Feb must be associated to the previous year if water year is < 3. I changed the function 'get_waterYear' accordingly`:

```
get_waterYear_new <- function(x, wyMonth=10L, numeric=TRUE) {
  x <- as.POSIXlt(x)
  yr <- x$year + 1900L
  mn <- x$mon + 1L
  if (wyMonth < 3){
    yr <- yr - ifelse(mn < wyMonth, 1L, 0L)
  }else{
    yr <- yr + ifelse(mn < wyMonth, 0L, 1L) 
  }
  if(numeric)
    return(yr)
  ordered(yr)
}
```
and ran a test by counting the number of days per year for all possible 12 wyMonth definitions:

```
library(EflowStats)
x <- data.frame(
  date = seq.Date(from = as.Date('1999-01-01'), to = as.Date('2003-12-31'), by = 'days')
)
for (wy_month in 1:12){
  cat("wyMonth =", wy_month, "NEW:\n")
  print(as.data.frame(table(get_waterYear_new(x$date, wy_month))))
  cat("wyMonth =", wy_month, "OLD:\n")
  print(as.data.frame(table(get_waterYear(x$date, wy_month))))
  cat("----------------\n\n")
}
```
Output:
```
wyMonth = 1 NEW:
  Var1 Freq
1 1999  365
2 2000  366
3 2001  365
4 2002  365
5 2003  365
wyMonth = 1 OLD:
  Var1 Freq
1 2000  365
2 2001  366
3 2002  365
4 2003  365
5 2004  365
----------------

wyMonth = 2 NEW:
  Var1 Freq
1 1998   31
2 1999  365
3 2000  366
4 2001  365
5 2002  365
6 2003  334
wyMonth = 2 OLD:
  Var1 Freq
1 1999   31
2 2000  365
3 2001  366
4 2002  365
5 2003  365
6 2004  334
----------------

wyMonth = 3 NEW:
  Var1 Freq
1 1999   59
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  306
wyMonth = 3 OLD:
  Var1 Freq
1 1999   59
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  306
----------------

wyMonth = 4 NEW:
  Var1 Freq
1 1999   90
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  275
wyMonth = 4 OLD:
  Var1 Freq
1 1999   90
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  275
----------------

wyMonth = 5 NEW:
  Var1 Freq
1 1999  120
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  245
wyMonth = 5 OLD:
  Var1 Freq
1 1999  120
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  245
----------------

wyMonth = 6 NEW:
  Var1 Freq
1 1999  151
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  214
wyMonth = 6 OLD:
  Var1 Freq
1 1999  151
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  214
----------------

wyMonth = 7 NEW:
  Var1 Freq
1 1999  181
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  184
wyMonth = 7 OLD:
  Var1 Freq
1 1999  181
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  184
----------------

wyMonth = 8 NEW:
  Var1 Freq
1 1999  212
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  153
wyMonth = 8 OLD:
  Var1 Freq
1 1999  212
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  153
----------------

wyMonth = 9 NEW:
  Var1 Freq
1 1999  243
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  122
wyMonth = 9 OLD:
  Var1 Freq
1 1999  243
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  122
----------------

wyMonth = 10 NEW:
  Var1 Freq
1 1999  273
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   92
wyMonth = 10 OLD:
  Var1 Freq
1 1999  273
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   92
----------------

wyMonth = 11 NEW:
  Var1 Freq
1 1999  304
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   61
wyMonth = 11 OLD:
  Var1 Freq
1 1999  304
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   61
----------------

wyMonth = 12 NEW:
  Var1 Freq
1 1999  334
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   31
wyMonth = 12 OLD:
  Var1 Freq
1 1999  334
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   31
----------------
```

(The problem described here https://github.com/DOI-USGS/EflowStats/issues/193 might be related to this issue).

**Comment by dblodgett-usgs** [comment]
Hi @jenskiesel -- sorry, but there isn't direct project support for EflowStats right now, so attention is pretty limited. 

I would be happy to take a pull request to fix this is you want to make the changes in a fork. 

@jds485 can you help with review?

**Comment by jenskiesel** [comment]
Thanks, @dblodgett-usgs, I have updated the function in my fork and created a new pull request.
@jds485, thanks if you can look into this and let me know if you have any questions.

**Comment by jds485** [comment]
Sorry I missed this earlier in the week - the notification went to my personal email. 

I looked over the edit and I think the original code does what is supposed to happen. For example, for water years that begin in Feb., WY 2000 would be Feb. 1999 - Jan. 2000, no leap day in WY 2000. That is what the original code does. It might be confusing to see Feb. 2000 be counted in WY 2001, thus leading to WY 2001 having a leap day, but that is correct in my opinion. What do you think @jenskiesel?

**Comment by jenskiesel** [comment]
Thanks @jds485 for looking into this. 

I agree that the assignment of WY is then not consistent. But the problem is, if WY2000 would be Feb.1999-Jan.2000, it has only 365 days. And then the "validate_data" function fails on the leap year test. Also, see the problem in the printout in my first message if the WY should begin in Jan (usual calendar year) then the original function assigns the wrong year (2000 for 1999).

I think for the cases where users need the water year to begin in Jan or Feb, we either need to change the assignemnt of WY (my changes to the code) or we need to modify the "validate_data" function?

**Comment by jds485** [comment]
> And then the "validate_data" function fails on the leap year test

Okay, thanks. I did not realize that this was throwing an error. Some edit will be needed.

> if the WY should begin in Jan (usual calendar year) then the original function assigns the wrong year (2000 for 1999)

The currently assigned year is consistent with using Feb. - Dec. as the starting month, but I agree it's confusing. The issue seems to be one of preference for how to assign a water year for different start months. I would argue that the water year value should be the calendar year with the most dates in the water year. So, for your example starting in 1999-01-01, water years for Jan. - June starting months would be assigned WY 1999, and July - Dec. would be WY 2000. I'm not sure of the convention, I'm just thinking about how someone might want to think about WY relative to calendar years. What do you think?

So, that would lead to the following edits:
1. Fix the water year assignment when using Jan - June as the start year, and describe in the documentation
2. Change the `validate_data` function to handle these edits to the WY definition
3. Check compatibility with other functions (hopefully the tests will catch any issues)

**Comment by jenskiesel** [comment]
Thanks for these good thoughts, @jds485.

> the water year value should be the calendar year with the most dates in the water year

I agree that your suggestion is more intuitive than using March as the cutoff. I did a quick search but could not find a convention/guideline on the correct assignment. 
 
Anyhow, if I implement:
```
if (wyMonth < 7){
```
instead of
```
if (wyMonth < 3){
```
in the "get_waterYear" function, we get the assignment of number of days per water year that I provided below.
We'd have to build in some "wyMonth" dependent checks into the "validate_data" function so that it does not fail for the months 3-6. So, I agree to your three action items and I can look into this (but I am currently on vacation until 26th of Aug).

```
wyMonth = 1 NEW:
  Var1 Freq
1 1999  365
2 2000  366
3 2001  365
4 2002  365
5 2003  365
wyMonth = 1 OLD:
  Var1 Freq
1 2000  365
2 2001  366
3 2002  365
4 2003  365
5 2004  365
----------------

wyMonth = 2 NEW:
  Var1 Freq
1 1998   31
2 1999  365
3 2000  366
4 2001  365
5 2002  365
6 2003  334
wyMonth = 2 OLD:
  Var1 Freq
1 1999   31
2 2000  365
3 2001  366
4 2002  365
5 2003  365
6 2004  334
----------------

wyMonth = 3 NEW:
  Var1 Freq
1 1998   59
2 1999  366
3 2000  365
4 2001  365
5 2002  365
6 2003  306
wyMonth = 3 OLD:
  Var1 Freq
1 1999   59
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  306
----------------

wyMonth = 4 NEW:
  Var1 Freq
1 1998   90
2 1999  366
3 2000  365
4 2001  365
5 2002  365
6 2003  275
wyMonth = 4 OLD:
  Var1 Freq
1 1999   90
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  275
----------------

wyMonth = 5 NEW:
  Var1 Freq
1 1998  120
2 1999  366
3 2000  365
4 2001  365
5 2002  365
6 2003  245
wyMonth = 5 OLD:
  Var1 Freq
1 1999  120
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  245
----------------

wyMonth = 6 NEW:
  Var1 Freq
1 1998  151
2 1999  366
3 2000  365
4 2001  365
5 2002  365
6 2003  214
wyMonth = 6 OLD:
  Var1 Freq
1 1999  151
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  214
----------------

wyMonth = 7 NEW:
  Var1 Freq
1 1999  181
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  184
wyMonth = 7 OLD:
  Var1 Freq
1 1999  181
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  184
----------------

wyMonth = 8 NEW:
  Var1 Freq
1 1999  212
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  153
wyMonth = 8 OLD:
  Var1 Freq
1 1999  212
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  153
----------------

wyMonth = 9 NEW:
  Var1 Freq
1 1999  243
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  122
wyMonth = 9 OLD:
  Var1 Freq
1 1999  243
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004  122
----------------

wyMonth = 10 NEW:
  Var1 Freq
1 1999  273
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   92
wyMonth = 10 OLD:
  Var1 Freq
1 1999  273
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   92
----------------

wyMonth = 11 NEW:
  Var1 Freq
1 1999  304
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   61
wyMonth = 11 OLD:
  Var1 Freq
1 1999  304
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   61
----------------

wyMonth = 12 NEW:
  Var1 Freq
1 1999  334
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   31
wyMonth = 12 OLD:
  Var1 Freq
1 1999  334
2 2000  366
3 2001  365
4 2002  365
5 2003  365
6 2004   31
----------------
```

**Comment by jds485** [comment]
That sounds great - thanks!

**Comment by jenskiesel** [comment]
@jds485, I implemented the updates we discussed in the [pull request](https://github.com/DOI-USGS/EflowStats/pull/197). Looking forward to your thoughts on that. Thanks!

**Comment by dblodgett-usgs** [comment]
Fixed in #197 and #198
